hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -533,10 +537,103 @@ def should_generate_more():
+        # cases for an initial period, by restriscting ourselves to test cases
+        # that are not much larger than it.
+        #
+        # Calculating the actual minimal generated test case is hard, so we
+        # take a best guess that zero extending a prefix produces the minimal
+        # test case starting with that prefix (this is true for our built in
+        # strategies). This is only a reasonable thing to do if the resulting
+        # test case is valid. If we regularly run into situations where it is
+        # not valid then this strategy is a waste of time, so we want to
+        # abandon it early. In order to do this we track how many times in a
+        # row it has failed to work, and abort small test case generation when
+        # it has failed too many times in a row.
+        consecutive_zero_extend_is_invalid = 0
+
         while should_generate_more():
             prefix = self.generate_novel_prefix()
+            assert len(prefix) <= BUFFER_SIZE
+
+            # We control growth during initial example generation, for two
+            # reasons:
+            #
+            # * It gives us an opportunity to find small examples early, which
+            #   gives us a fast path for easy to find bugs.
+            # * It avoids low probability events where we might end up
+            #   generating very large examples during health checks, which
+            #   on slower machines can trigger HealthCheck.too_slow.
+            #
+            # The heuristic we use is that we attempt to estimate the smallest
+            # extension of this prefix, and limit the size to no more than
+            # an order of magnitude larger than that. If we fail to estimate
+            # the size accurately, we skip over this prefix and try again.
+            #
+            # We need to tune the example size based on the initial prefix,
+            # because any fixed size might be too small, and any size based
+            # on the strategy in general can fall afoul of strategies that
+            # have very different sizes for different prefixes.
+            small_example_cap = clamp(10, self.settings.max_examples // 10, 50)
+
+            if (
+                self.valid_examples <= small_example_cap
+                and self.call_count <= 5 * small_example_cap
+                and not self.interesting_examples
+                and consecutive_zero_extend_is_invalid < 5
+            ):
+                minimal_example = self.cached_test_function(
+                    prefix + hbytes(BUFFER_SIZE - len(prefix))
+                )
+
+                if minimal_example.status < Status.VALID:
+                    consecutive_zero_extend_is_invalid += 1
+                    continue
+
+                consecutive_zero_extend_is_invalid = 0
+
+                max_length = (len(minimal_example.buffer) - len(prefix)) * 10 + len(
+                    prefix
+                )
```

Comment:
This sometimes produces values of `max_length` that are much larger than `BUFFER_SIZE` (e.g. 44001).

It looks like those values never end up in an actual test, but is there any reason not to clamp immediately, before consulting the tree?

Reply:
Also, this might be clearer as something like:
```python
smallest_extension = len(minimal_example.buffer) - len(prefix)
max_length = len(prefix) + (smallest_extension * 10)
```

That way it more obviously matches the comment up above.

Reply:
Good ideas both. Done.